### PR TITLE
Add support for @(InternalsVisibleTo) items that turn into assembly attributes

### LINKED ISSUE #3166: GenerateAssemblyInfo support for InternalsVisibleToAttribute and AssemblyMetadataAttribute
Hello,

In our .NET 4.7.2 applications we set the InternalsVisibleTo attribute on our libraries so that our internal unit test projects can run unit tests on internal methods without having to go through the trouble of putting public signatures of these methods into the class and wrapping everything in preprocessor directives.

We also make use of the AssemblyMetadata attribute to have our build system add vital diagnostic information to our assemblies so that there is never an issue with identifying when a project was built, from what repo commit etc.  This has made identifying and fixing bugs much easier more efficient.

To set these values we've manually set the InternalsVisibleTo attribute at design time into the AssemblyInfo.cs file.  We also have our build system inject the values into the AssemblyInfo.cs file at build time during the build (but the changes are not committed as the metadata all build version related metadata).

In .NET Standard and Core we've found that the common Assembly Attributes are now set in the project file.  Do you have plans to add these other attributes to the Microsoft.NET.GenerateAssemblyInfo.targets file so that they an be included in the project file so the new generate assembly info functionality can support adding these to the output assembly?

We have read there is a workaround where we can set the property GenerateAssemblyInfo to False and add an AssemblyInfo.cs file to the .NET Standard/Core projects but we would rather not use this workaround.  For one we do not know how long it will be supported and also we do not know if it will break the new project build workflow in some way.

Can you please provide some advice on how these attributes can be used in .NET Standard/Core project files without the workaround or an indication if you plan on adding support for them to the Microsoft.NET.GenerateAssemblyInfo.targets file?

Thank you,

Bryan

### PR DESCRIPTION
**Pull request by kzu**
Given that `InternalsVisibleTo` (IVT) is such a common assembly-level attribute, this adds support
for specifying it directly via simple items, such as:

```
  <ItemGroup>
    <InternalsVisibleTo Include="MyLibrary.Tests" />
  </ItemGroup>
```

Optionally, a `Key` metadata can be specified to provide a strong-named IVT:

```
  <ItemGroup>
    <InternalsVisibleTo Include="MyLibrary.Tests" Key="PUBLIC_KEY" />
  </ItemGroup>
```

The targets will also use automatically a `$(PublicKey)` if available and no `%(Key)` metadata
override is found. Otherwise, it will default to an IVT without a key.

This also avoids having to learn the `_Parameter1` syntax in `AssemblyAttribute` elements, and is
more similar to the way other higher-level properties like `AssemblyTitle` or `Product` are also
turned into assembly attributes.

To turn off this feature, set `$(GenerateInternalsVisibleToAttributes)` to `false`.

Partially fixes #3166

### CODE REVIEW

**Comment by peterhuene** [comment]
Hi @kzu.  This looks good.

Could you add some test coverage to this PR?

**Comment by nguerrera** [comment]
This is going to conflict with the unofficial feature in Arcade so would need to be coordinated.

https://github.com/dotnet/arcade/blob/671bfe2954fe2e0fc2052cb0ee59cbfe9e5e8801/src/Microsoft.DotNet.Arcade.Sdk/tools/GenerateInternalsVisibleTo.targets

That one supports using $(PublicKey) or %(Key) metadata AFAICT. I think we'll want this to be a superset of that in functionality, ideally. Or we have to be able to turn this one off to enable that one.

cc @tmat

**Comment by tmat** [comment]
We shouldn't end up with two ways of doing it. Let's have one impl that supports all we need.

**Comment by kzu** [comment]
Good point @nguerrera and @tmat. Added support for the existing arcade stuff too, which should make it a drop-in replacement. It can also be turned off now.

**Comment by kzu** [comment]
@peterhuene just like for the other PR, if you can point me to existing tests for `AssemblyAttribute` generation, I can add to that. Thanks!

**Comment by sharwell** [comment]
@nguerrera Where would a new feature like this get documented, such that once it's available we can point users to it?

**Comment by rainersigwald** [comment]
@sharwell https://docs.microsoft.com/en-us/visualstudio/msbuild/common-msbuild-project-items?view=vs-2019

**Comment by peterhuene** [comment]
@kzu [here is a test](https://github.com/dotnet/sdk/blob/master/src/Tests/Microsoft.NET.Build.Tests/GivenThatWeWantToControlGeneratedAssemblyInfo.cs#L36) that is setting the individual properties that cause additional `AssemblyAttribute` items to be defined.

We should probably add a test to this file to cover setting `InternalsVisibleTo` items in the test project file and reflecting on the output assembly to ensure the attributes were set.

**Comment by kzu** [comment]
`AssemblyAttribute` is already missing from docs :(

**Comment by kzu** [comment]
@peterhuene finally got to implement the unit tests, they cover (I think) all the cases supported by the targets.

**Comment by kzu** [comment]
Added docs PR too :)

**Comment by kzu** [comment]
Was that fail a fluke?

**Comment by kzu** [comment]
Is there something I can do to get this PR green?

**Comment by dsplaisted** [comment]
> Is there something I can do to get this PR green?

One leg was failing but the results were no longer available, so I've restarted CI

File: src/Tests/Microsoft.NET.Build.Tests/GivenThatWeWantToControlGeneratedAssemblyInfo.cs [hunk 1]
```diff
@@ -349,6 +350,119 @@ BuildCommand BuildProject(string buildNumber)
...
+                {
+                    var ns = project.Root.Name.Namespace;
+
+                    project.Root.Add(
+                        new XElement(ns + "ItemGroup",
+                            new XElement(ns + "InternalsVisibleTo",
+                                new XAttribute("Include", "Tests"))));
+                });
+
+            new RestoreCommand(Log, testAsset.TestRoot).Execute().Should().Pass();
```
Comment (sfoslund) [inline src/Tests/Microsoft.NET.Build.Tests/GivenThatWeWantToControlGeneratedAssemblyInfo.cs]: nit: Same as #3440, restore is no longer necessary

**Review by sfoslund — APPROVED** [review]
This is green now and it looks good to me. @dsplaisted is this ready to go?

**Comment by dsplaisted** [comment]
Well, I merged #3440, and now this PR has conflicts that need to be resolved.  @sfoslund or @kzu, could you take care of that?  And while you're doing that it would be good to remove the unnecessary restore calls from the tests both in this PR and in the tests added in #3440.

**Comment by sfoslund** [comment]
@dsplaisted @kzu this is merged and ready to go, as far as I can tell.

**Comment by DanVicarel** [comment]
Any idea in what version of VS we can expect to see this change?

**Comment by ggirard07** [comment]
@DanVicarel as this seems to be associated with milestone 5.0.1, I won't expect it prior next .Net Core major release [(which is currently planed to be re-branded to .Net 5](https://devblogs.microsoft.com/dotnet/introducing-net-5/)), then next Visual Studio update to integrate it...

**Comment by renatfx** [comment]
@kzu probably it's already late to ask/comment, but I would like to

Why did you name item metadata `Key` and not `PublicKey`?
Finally it goes to IVT contructor as assembly name PublicKey which it is

Why did you introduce PublicKey property?
With the ItemDefinitionGroup feature of MSBuild we can also set the default public key but, what I personally prefer, limiting it to certain items only (comparing to property):
```xml
<ItemDefinitionGroup>
  <InternalsVisibleTo>
    <PublicKey>...</PublicKey>
  </InternalsVisibleTo>
</ItemDefinitionGroup>
```

**Comment by kzu** [comment]
@renatfx it's never late to ask good questions :). The item metadata was named `Key` for compatibility with the [Arcade SDK](https://github.com/dotnet/arcade/blob/master/src/Microsoft.DotNet.Arcade.Sdk/tools/GenerateInternalsVisibleTo.targets) specifically. Likewise for the `$(PublicKey)` property.

Finally, doing this in an item definition group is something customers can do themselves. Granted, the SDK could provide that item definition group conditionally, like:

```
<ItemDefinitionGroup Condition="'$(PublicKey)' != ''">
  <InternalsVisibleTo>
    <PublicKey>$(PublicKey)</PublicKey>
  </InternalsVisibleTo>
</ItemDefinitionGroup>
```

But it's not entirely clear that every time the current project is strong named, its IVT will point to 
other projects with the same key. They might have a different one, and now you'd end up with 
a default behavior to turn off...

**Comment by renatfx** [comment]
Well, for me as a .NET (SDK) consumer rather than developer, this is asbolutely unclear:

_Why should public .NET SDK be compatible wth internal Arcade SDK?_

btw, I didn't realize that `$(PublicKey)` property has public key of the current project as its value... now I see that using it as an assembly name public key in IVT .ctor by default, only when the current assembly is signed itself, defenitely makes sense.

However, I still would vote (if I could) for `PublicKey` rather than `Key` metadata item name.

**Comment by tmat** [comment]
> Why should public .NET SDK be compatible wth internal Arcade SDK?

It does not. We can rename the `Key` metadata to `PublicKey` if it fits better. I filed https://github.com/dotnet/arcade/issues/4658 to follow up in Arcade to use the new official feature, where we can also rename it to follow to whatever name the .NET SDK decides to use.

**Comment by renatfx** [comment]
@tmat ok, great!

**Comment by shueybubbles** [comment]
I see in the binlog that the right public key is being passed in the target, but the emitted assemblyinfo.cs is missing the key. I have VS 17.10.4 and .Net 8 304 SDK. What might cause the target to emit the attribute without the key? 
image

**Pull request merged**